pageserver/src/waldecoder.rs

```diff
@@ -104,11 +75,15 @@ impl WalStreamDecoder {
             if self.lsn.segment_offset(pg_constants::WAL_SEGMENT_SIZE) == 0 {
                 // parse long header
 
-                if self.inputbuf.remaining() < SizeOfXLogLongPHD {
+                if self.inputbuf.remaining() < XLOG_SIZE_OF_XLOG_LONG_PHD {
                     return Ok(None);
                 }
 
-                let hdr = self.decode_XLogLongPageHeaderData();
+                let mut longhdrbuf: BytesMut = BytesMut::new();
+                longhdrbuf.reserve(XLOG_SIZE_OF_XLOG_LONG_PHD);
+                longhdrbuf = self.inputbuf.split_to(XLOG_SIZE_OF_XLOG_LONG_PHD);
+
+                let hdr = XLogLongPageHeaderData::from_bytes(&mut longhdrbuf.freeze());
```

Comment:
`BytesMut` is overkill here. We should be able to decode from a slice `&[u8]`.

Also, `BytesMut::split_to` returns a new `BytesMut` by value, so the first `BytesMut` you are creating is thrown away.

Ideally, these four lines should be reduced to just:
```
let hdr = XLogLongPageHeaderData::from_bytes(&self.inputbuf);
```

Reply:
Thank you for the review.

I've changed the code to manual parsing from slice.
The new version is as weird as the previous one, but at least it seems to be a safe option until we agree on transmute/zerocopy solution for decoding.

Reply:
Sorry, looks like I gave bad feedback-- I didn't realize that `inputbuf.advance` would be needed.

`&[u8]` implements the `Buf` interface, but it's a little harder than I thought to support plain slice and `Bytes` sources simultaneously. We could either put it back the way it was, or do something like:
```rust
impl XLogPageHeaderData {
    pub fn from_bytes<B: Buf>(buf: &mut B) -> XLogPageHeaderData {
        XLogPageHeaderData {
            xlp_magic: buf.get_u16_le(),
            xlp_info: buf.get_u16_le(),
            xlp_tli: buf.get_u32_le(),
            xlp_pageaddr: buf.get_u64_le(),
            xlp_rem_len: buf.get_u32_le(),
            padding: buf.get_u32_le(),
        }
    }
}
```
That's almost the same as what you started with, and its support for plain slices is kind of awkward:
```rust
// This is a &[u8].
let mut my_slice = &some_data[..];
// This "advances" by moving the reference forward.
let hdr = XLogPageHeaderData::from_bytes(&mut my_slice);
```

I'm not sure if that extra capability is worth the trouble, though.

Reply:
Well. The final goal is to reuse this code in `find_end_of_wal_segment` which now parses the same fields in an ad hoc way and uses `byteorder` functions to work with a slice.

Eventually, we should have only one function (set of functions) to parse wal files. And both `find_end_of_wal_segment` and `poll_decode` should use it.

To be honest, I don't see a significant difference between different ways of implementing this parsing. So I'll take your word for it.